tests/agent_features/test_custom_events.py

```diff
@@ -229,3 +231,103 @@ def test_transaction_create_custom_event_not_called():
 def test_application_create_custom_event_not_called():
     app = application()
     record_custom_event("FooEvent", _user_params, application=app)
+
+
+# Test completness of LLM content/input despite attribute limits being set
+
+
+LlmChatCompletionMessage = {
+    "content": "A" * 9001,  # Should print out to completion
+    "input": "B" * 5000,  # Should cap out at 255 or 4095
+    "foo": "b" + "a" * 6000 + "r",
+}
+LlmEmbedding = {
+    "content": "A" * 9001,  # Should cap out at 255 or 4095
+    "input": "B" * 5000,  # Should print out to completion
+    "foo": "b" + "a" * 6000 + "r",
+}
+SomeOtherDict = {
+    "content": "A" * 9001,
+    "input": "B" * 5000,
+    "foo": "b" + "a" * 6000 + "r",
+}
+
+
+recorded_events_256 = [
+    (
+        {"type": "LlmChatCompletionMessage"},
+        {
+            "content": "A" * 9001,  # Should print out to completion
+            "input": "B" * 255,  # Should cap out at 255 or 4095
+            "foo": "b" + "a" * 254,
+        },
+    ),
+    (
+        {"type": "LlmEmbedding"},
+        {
+            "content": "A" * 255,  # Should cap out at 255 or 4095
+            "input": "B" * 5000,  # Should print out to completion
+            "foo": "b" + "a" * 254,
+        },
+    ),
+    (
+        {"type": "SomeOtherDict"},
+        {
+            "content": "A" * 255,
+            "input": "B" * 255,
+            "foo": "b" + "a" * 254,
+        },
+    ),
+]
+
+recorded_events_4096 = [
+    (
+        {"type": "LlmChatCompletionMessage"},
+        {
+            "content": "A" * 9001,  # Should print out to completion
+            "input": "B" * 4095,  # Should cap out at 255 or 4095
+            "foo": "b" + "a" * 4094,
+        },
+    ),
+    (
+        {"type": "LlmEmbedding"},
+        {
+            "content": "A" * 4095,  # Should cap out at 255 or 4095
+            "input": "B" * 5000,  # Should print out to completion
+            "foo": "b" + "a" * 4094,
+        },
+    ),
+    (
+        {"type": "SomeOtherDict"},
+        {
+            "content": "A" * 4095,
+            "input": "B" * 4095,
+            "foo": "b" + "a" * 4094,
+        },
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "max_val,expected_events",
+    (
+        (255, recorded_events_256),
+        (4095, recorded_events_4096),
+    ),
+)
+def test_create_custom_event_no_limit(max_val, expected_events):
```

Comment:
Let's take full advantage or parametrize here and do something like the code below instead. I find this more explicit and concise as it has both the input and expected output next to each other. This way we aren't testing multiple events at once as well-which is generally considered better practice. There's that quote about, "Tests should test one and one thing only." Admittedly this is something we do very poorly in our code base today but we can still try to move in the right direction when writing new tests. Given that the 255 or 4095 is just the max attribute setting, I don't think we need to test the full min and max range here. We have other tests to make sure the setting is set correctly based on it's allowed range. Maybe just pick a random value for max attribute like 300 and roll with it. Also note I dropped the if transaction from this. There should always be a transaction because you are using the background_task decorator.
```suggestion
@pytest.mark.parametrize(
    "event_type,event_data,expected_event_data",
    (
        [
            "LlmChatCompletionMessage",
            {
                "content": "A" * 9000,
                "input": "B" * 9000,
                "foo": "b"* 9000,
            },
            {
                "content": "A" * 9000,
                "input": "B" * 300,
                "foo": "b" * 300,
            },
        ],
        [
            "LlmEmbedding",
            {
                "content": "A" * 5001,
                "input": "B" * 5001,
                "foo": "b"* 5001,
            },
            {
                "content": "A" * 300,
                "input": "B" * 5001,
                "foo": "b" * 300,
            },
        ],
        [
            "MyCustomEvent",
            {
                "content": "A" * 5001,
                "input": "B" * 5001,
                "foo": "b"* 5001,
            },
            {
                "content": "A" * 300,
                "input": "B" * 300,
                "foo": "b" * 300,
            },
        ],
    ),
)
def test_create_custom_event_no_limit(event_type, event_data, expected_event_data):
    @reset_core_stats_engine()
    @override_application_settings({"custom_insights_events.max_attribute_value": 300})
    @validate_custom_event_count(3)
    @validate_custom_events(({"type": event_type}, expected_event_data))
    @background_task()
    def _test():
        transaction.record_custom_event(event_type, event_body)

    _test()

```